### PR TITLE
Feature/iban attribute

### PR DESCRIPTION
**Pull request by deepweeb**
**Solution for Issue #29**

IBAN attribute is fully implemented with all tests passing.
We now have one more field as input when adding a contact (next to the senderhash and sender pubkey).
Also conversion is edited so that, it also includes sequence number of the lastest block+1 and the IBAN for calculation of the hash.

### CODE REVIEW

File: app/src/main/java/nl/tudelft/b_b_w/controller/ConversionController.java [hunk 1]
```diff
@@ -8,21 +8,25 @@
  */
 public class ConversionController {
     // Variables which we need to create a hashed key
-    private String keyOwner;
+    private String blockOwner;
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/b_b_w/controller/ConversionController.java]: Why did you change `keyOwner `into `blockOwner`

Comment (deepweeb) [inline app/src/main/java/nl/tudelft/b_b_w/controller/ConversionController.java]: Because when you use keyOwner you are refering to the PublicKeyOwner. This is totally not correct. This should be changed to blockOwner to prevent confusion.

**Comment by nzarin** [comment]
Do all tests pass?

File: app/src/main/java/nl/tudelft/b_b_w/controller/ConversionController.java (hunk 1 above)
Comment (ashay123) [inline app/src/main/java/nl/tudelft/b_b_w/controller/ConversionController.java]: Great!

**Comment by deepweeb** [comment]
Yes all tests passed as said in the commit

**Comment by nzarin** [comment]
Cool, looks fine.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 2]
```diff
@@ -84,7 +84,7 @@ public BlockController(Context _context) {
         String owner = block.getOwner();
         Block newBlock = BlockFactory.getBlock("REVOKE", block.getOwner(), block.getSequenceNumber(),
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: You should not be able to give a sequence number upon creating a block.

File: app/src/main/java/nl/tudelft/b_b_w/model/Block.java [hunk 3]
```diff
@@ -21,16 +22,18 @@
      * @param _ownHash            our own hash
      * @param _previousHashChain  the hash value of the block before in the chain
      * @param _previousHashSender the hash value of the block before of the sender
-     * @param _publicKey          public key of the owner of the block
+     * @param _publicKey          public key of the sender
+     * @param _iban               IBAN number of a contact
      * @param _isRevoked          boolean to check whether a block is revoked or not
      */
-    public Block(String _owner, int _sequenceNumber, String _ownHash, String _previousHashChain, String _previousHashSender, String _publicKey, boolean _isRevoked) {
+    public Block(String _owner, int _sequenceNumber, String _ownHash, String _previousHashChain, String _previousHashSender, String _publicKey, String _iban, boolean _isRevoked) {
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/model/Block.java]: You should not be able to give a sequencenumber to the constructor upon creating a block

File: app/src/main/java/nl/tudelft/b_b_w/model/BlockFactory.java [hunk 4]
```diff
@@ -21,16 +21,17 @@
...
      * @param _previousHashSender Previous hash value of the sender of the block
-     * @param _publicKey          Public key of the block
+     * @param _publicKey          Public key of the sender
+     * @param _iban               IBAN number of a contact
      * @return Returns a new block
      * @throws IllegalArgumentException when the input string type is invalid
      */
-    public static Block getBlock(String type, String _owner, int _sequenceNumber, String _ownHash, String _previousHashChain, String _previousHashSender, String _publicKey) throws IllegalArgumentException {
+    public static Block getBlock(String type, String _owner, int _sequenceNumber, String _ownHash, String _previousHashChain, String _previousHashSender, String _publicKey, String _iban) throws IllegalArgumentException {
         switch (type) {
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/model/BlockFactory.java]: You should not be able to give a sequencenumber to the constructor upon creating a block

File: app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java [hunk 5]
```diff
@@ -70,11 +70,15 @@ public void onAddBlock(View view) {
...
+
 
             // create and add the block
             Block previous = handler.getLatestBlock(ownerName);
-
-            ConversionController conversionController = new ConversionController(ownerName, senderPublicKey, previous.getOwnHash(), senderHash);
+            int blockSeqNumber = previous.getSequenceNumber()+ 1;
+            ConversionController conversionController = new ConversionController(ownerName, blockSeqNumber, senderPublicKey, previous.getOwnHash(), senderHash, senderIban);
             String ownHash = conversionController.hashKey();
             Block block = new Block(
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java]: You should not create a block this way, but using the blockcontroller

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 6]
```diff
@@ -52,7 +53,7 @@
     public void setUp() {
         this.bc = new BlockController(RuntimeEnvironment.application);
         this._block = BlockFactory.getBlock(TYPE_BLOCK, owner, sequenceNumber, ownHash,
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: You should not be able to give a sequencenumber to the constructor upon creating a block

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 7]
```diff
@@ -75,7 +76,7 @@ public void testAddBlock() throws Exception {
     public void testAddBlock2() throws Exception {
         String newOwner = owner+"2";
         Block newBlock = BlockFactory.getBlock(TYPE_BLOCK, newOwner, sequenceNumber, ownHash,
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: You should not be able to give a sequencenumber to the constructor upon creating a block

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 8]
```diff
@@ -103,7 +104,7 @@ public void testAddDupBlocks() {
     @Test(expected=RuntimeException.class)
     public void alreadyRevoked() {
         Block newBlock = BlockFactory.getBlock(TYPE_REVOKE, owner, sequenceNumber, ownHash,
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: You should not be able to give a sequencenumber to the constructor upon creating a block

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 9]
```diff
@@ -115,7 +116,7 @@ public void alreadyRevoked() {
     public void testEmptyList() {
         bc.addBlock(_block);
         Block newBlock = BlockFactory.getBlock(TYPE_BLOCK, owner, sequenceNumber+1, ownHash,
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: You should not be able to give a sequencenumber to the constructor upon creating a block

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java [hunk 10]
```diff
@@ -33,7 +34,7 @@
      */
     @Before
     public void makeNewBlock() throws Exception {
-        _block = new Block(owner, sequenceNumber, ownHash, previousHashChain, previousHashSender, publicKey, isRevoked);
+        _block = new Block(owner, sequenceNumber, ownHash, previousHashChain, previousHashSender, publicKey,iban, isRevoked);
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java]: You should not create a block this way.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 2 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: I did not write BlockFactory or this line of code. Should be a separate issue

File: app/src/main/java/nl/tudelft/b_b_w/model/Block.java (hunk 3 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/b_b_w/model/Block.java]: This is a separate issue, will fix in another pull request.

File: app/src/main/java/nl/tudelft/b_b_w/model/BlockFactory.java (hunk 4 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/b_b_w/model/BlockFactory.java]: Same story

File: app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java (hunk 5 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java]: See #15

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java (hunk 10 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java]: I did not write BlockFactoryUnitTest. I cannot just change the whole method (which will result in changing other methods and me working on other issue). Someone whos responsible for this class (@ashay123 toch?) should be making an issue and then fix this.

**Comment by deepweeb** [comment]
If you review the code, please only review the changed parts of the code, create another issue if you spot problems that is NOT related to the pull request, read the pull request description. Thanks

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java (hunk 10 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java]: new issue and hotfix i is created to address this problem

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 2 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: According to git blame, you did write this code.

File: app/src/main/java/nl/tudelft/b_b_w/model/Block.java (hunk 3 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/model/Block.java]: Since you are editing the block attributes now already, you can easily fix this issue too.

File: app/src/main/java/nl/tudelft/b_b_w/model/BlockFactory.java (hunk 4 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/model/BlockFactory.java]: Since you are editing the block attributes now already, you can easily fix this issue too.

File: app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java (hunk 5 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java]: Since you are editing the block attributes now already, you can easily fix this issue too.

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java (hunk 10 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockFactoryUnitTest.java]: Next to that, you missed a space left of 'iban'

**Comment by humanjas** [comment]
If you do are modifiying the code of the block, then you should be able to fix it as well.

**Comment by deepweeb** [comment]
What is "should". As I said, if I fix the issue that's not related to the issue this pull request is for, this fix might interfere with another methods which might lead  to a long fix (which should be a separate issue.)

**Comment by humanjas** [comment]
Since you are editing the constructor of the block class, next to the tests that correspond to it, there is no extra possibility of breaking the code if you remove the sequenceNumber attribute.

File: app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java (hunk 5 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java]: #15 is a whole another big issue, theres no quick fix. It should be fixed all at once as this kind of error are in another files that I have not touched too.

**Comment by deepweeb** [comment]
This error occurs in another files that is not being edited as well. If i changes it here: 1 it does not related to my issue, 2 it might interfere with another issue (in this example not), 3 it means that I only fix the issue in the files that i edit. 
Please work according to the guidelines, read the issue, review the changes solution TO that issue. 
The issue mentioned will be fixed in #15 if it's small.

**Comment by humanjas** [comment]
You edited all files, where it occurs in. Next to that, a pull request does not need to relate to a single issue. You have edited all occurences of the initialization of the block. These are the same occurences as the occurences as the sequence number. Since your changes to the files still contain bugs, I cannot approve it.

**Comment by ashay123** [comment]
@deepweeb Even though I agree with your arguments(well, most parts), I think that @humanjas also makes a valid point. You made changes in some files, and kept the style of the code, in this case making a new block, the same. However as you know this is the wrong way for making a block, it is easier to do it right immediately, regardless of other issues that are still pending. 
As for now I see this PR as incomplete, please make these changes that are pointed out.

**Comment by deepweeb** [comment]
Let me ask Stefan, if I should fix a bug that is in a whole project with a pull request that is not meant to fix this bug.
To make this workflow clear, the pull request should be approved and another issue should be made and be fixed with another issue.
It's the same as if you guys make a pull request, and then having me asking you to fix a project wide bug from the code (that could even be in the files that you have not edited) that is not related to the issue just because I think it is "convenience" (it is not from my point of view). Since this pull request is a solution to the mentioned issue #29, approve this one so I can create another issue regarding this. Remember, we are in the feature/IBAN_Attribute, a complete feature branch, not bug fix or merge branch.

**Comment by nzarin** [comment]
Guys, I get your points. I think the problem is mainly because we did not do this before (fix everything before merging, that is why there are still bugs). There might be more than one person working on that branch. So everyone who took part of writing these bugs must also fix them, not just one person. For this scenario, in my opinion, a separate issue must be made to fix these bugs. After everything is fixed , then we should work in the style Jasper suggested. It's still good that you guys said it and put attention to it.

**Comment by deepweeb** [comment]
Since I have minimum approvals (2 approvals from the group), and I do need to merge so I can work on the next issue that needs this fix. Besides this is not something catastrophic and could be addressed in another issue.  I will go ahead and merge.

**Comment by deepweeb** [comment]
Issue generated for the problems: #46

File: app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java (hunk 5 above)
Comment (ashay123) [inline app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java]: The creation of blocks does not happen in DatabaseHandler, but BlockFactory. As @humanjas pointed out, the blocks are instantiated using the Block class, but this should happen in BlockFactory. So this issue isn't the same as stated in #15.

Comment (deepweeb) [inline app/src/main/java/nl/tudelft/b_b_w/view/AddBlockActivity.java]: #46
